### ISSUE TITLE
Add support for nested component validation

### ISSUE THREAD

**Issue by dixhuit** [victorgarciaesgi/regle#40]
### What problem is this solving

Vuelidate supported nested component validation so this would go some way towards helping Vuelidate users that relied on this feature migrate to Regle.

The suggestion stems from this conversation with the Regle project maintainer on the following Vuelidate GitHUb issue: https://github.com/vuelidate/vuelidate/issues/1163#issuecomment-2566090785

### Proposed solution

Groups of form inputs could be provided by child components which may or may not be included in the greater form (example use case is a highly dynamic form where the user makes choices that influence what groups of inputs are present). These child components declare their own validation rules making for a modular approach. On testing validation state, child components validate themselves and handle their own validation messaging, the parent form is made aware of whether the entire form is valid or invalid.

**Comment by martinszeltins** [comment]
Couldn't this be done using Pinia store already? Not sure how this fits into the data-driven approach of Regle. Would like to see some real world examples of this.

**Comment by victorgarciaesgi** [comment]
Do you have by chance a simple reproduction exemple of what it looked like in Vuelidate? And what compromise you need to have?
- Do the children always have the same parent? 
- Are the children completly isolated from the rest of the app logic
- Are the children functionnal components? (only depending on props)

Also joining @martinszeltins , what is preventing you to use a Pinia store to have a shared validation state? (Apart from the migration convenience)

**Comment by vad1ym** [comment]
+1, I want to migrate from Vuelidate to Regle, but I'm using nested validation.

For example, I have a number of fields (components), each with its own specific validation, which is configured via props. These fields are rendered from a service-provided array of field definitions.

I want to be able to validate all these fields from the parent component with a single click, as is done in Vuelidate, without exposing the r$ from each component or using a global state.

**Comment by victorgarciaesgi** [comment]
If there is a need for it I will work on it. But I also fear the performance side as in another thread the Vuelidate maintainer said their solution is not ideal.
If you just have a quick Stackblitz exemple at hand @vad1ym that would help because i'm not used to this feature.
To avoid loosing the type safety, I guess only the root nested results will be stored in a `$nestedResults` property.

**Comment by NtchPlayer** [comment]
> +1, I want to migrate from Vuelidate to Regle, but I'm using nested validation.
> 
> For example, I have a number of independent fields (components), each with its own specific validation, which is configured via props. These fields are rendered from a service-provided array of field definitions.
> 
> I want to be able to validate all these fields from the parent component with a single click, as is done in [Vuelidate](https://vuelidate-next.netlify.app/advanced_usage.html#nested-validations), without exposing the r$ from each component or using a global state.
> 
> Also, it adds another approach to implementing collections (via nested components instead of $each)

I am exactly in the same case as yours. Regle seems to be a very good alternative of Vuelidate, but my project is based on nested component validation. So I can't migrate to Regle without it. My logic of validation is encapsulate into my fields components. It's create very powerfull components. With that you can create forms without any effort to handle errors display (because they are manage directly into fields components).

I make a little reproduction of my app on stackblitz for you @victorgarciaesgi to see how it's work 😄 I really hope that you can't add this to Regle ! If you have any interrogation feel free to aks !

Reproduction: https://stackblitz.com/edit/vuelidate-nested-validation-example?file=app.vue

**Comment by victorgarciaesgi** [comment]
Thanks for the reproduction @NtchPlayer ! 👌 

I planned to add the nested component collection, and have some ideas to stay performant, because the actual implementation of Vuelidate will make no sense in my actual context.

What I plan is to have a `createSharedUseRegle` like this

```ts 
// [some-file.ts]
const { useCollectNestedValidations, useNestedRegle } = createNestedUseRegle()
```

```vue
<script setup lang='ts'>
// Child.vue

const {r$} = useNestedRegle()
</script>
```

```vue
<script setup lang='ts'>
// Parent.vue

const {r$} = useCollectNestedValidations()
</script>
```

Where `useNestedRegle` is a clone of `useRegle`, but can be collected by `useCollectNestedValidations` anywhere

**Comment by NtchPlayer** [comment]
That sound very cool !

Note that on vuelidate you can pass to some parameters to useVuelidate instance to config how nested validation work.

#### Scope validation process
You can add a scope to (that is very usefull if you have multiples forms on the same page).
```JS
const contactFormValidation = useVuelidate({ $scope: 'contact' });
```

I update my stackblitz if you what to see how it's work with vuelidate https://stackblitz.com/edit/vuelidate-nested-validation-example?file=app.vue

Vuelidate docs about $scope options: https://vuelidate-next.netlify.app/advanced_usage.html#scope-property

##### Stop propagation
You can add a paramater $stopPropagation to block the propagation of the validation to parent component. This is very usefull if on a page with a form and in this page you have a modal with a form too. With $stopPropagation you can encapsulate logic of the form that is in the modal (because the modal don't send her validation logic to the main form of the page).

Vuelidate docs about $stopPropagation (can be more clear than my explanation my english is so bad sorry 😢) : https://vuelidate-next.netlify.app/advanced_usage.html#stoppropagation-property

**Comment by victorgarciaesgi** [comment]
Thanks for the details and reproductions!
What's great is my idea of the implementation would support both `Scope validation` & `Stop propagation` out of the box, as using a different composable would make it opti-in

**Comment by victorgarciaesgi** [comment]
Hey @dixhuit , @vad1ym @NtchPlayer , released a beta version `0.7.4-beta.2` with support for scoped (validation).

Why **scoped**?  Because this allow to do nested component validation without really needing for it to be nested.

Current API looks exactly as what I commented before:

Define one or multiple scopes:

```ts
// scope.config.ts
import { createScopedUseRegle } from '@regle/core';

export const { useCollectScopedValidations, useScopedRegle } = createScopedUseRegle();

export const { useCollectScopedValidations: useCollectFooScopedValidations, useScopedRegle: useScopedFooRegle } = createScopedUseRegle();
```

Use the `useScopedRegle` in any Component

```vue
<script setup lang="ts">
// Child.vue
import { required } from '@regle/rules';
import { useScopedRegle } from './config';

const { r$ } = useScopedRegle({ name: '' }, { name: { required } });
</script>
```


Collect all nested/scoped validations with `useCollectScopedValidations`


```vue
<template>
     <Child/>
     <Child/>
</template>
<script setup lang="ts">
// Parent.vue
import Child from './Child.vue';
import { useCollectScopedValidations } from './config';

const { r$ } = useCollectScopedValidations();

</script>
```

I'm currently migrating your reproduction from Vuelidate to Regle

**Comment by victorgarciaesgi** [comment]
Here you go with the updated reproduction: https://stackblitz.com/edit/vuelidate-nested-validation-example-qj8c39t1?file=app.vue
(Don't know how you code without Typescript 😂 ), also spotted a SSR bug i'll fix in the same release

**Comment by victorgarciaesgi** [comment]
Some notes on this reproduction case:
- Case could have easily be done without nested collection, just props, i'll try to find a better use case exemple
  - For props, here is an exemple: https://reglejs.dev/typescript/typing-props#typing-a-field-prop 
- You don't need to define new validators to have custom messages, in Regle you can use [global configuration](https://reglejs.dev/advanced-usage/global-config)
  - `createScopedUseRegle` supports having a custom `useRegle` as optional parameter, keeping types and custom messages

```ts
const { useRegle: useCustomRegle } = defineRegleConfig({...});
export const { useCollectScopedValidations, useScopedRegle } = createScopedUseRegle(useCustomRegle);
```

Glad I could come up with a solution for all of you! While keeping the Regle API clean and types 👍 

Don't hesitate to try the `0.7.4-beta.3` when released (will publish a SSR fix) and tell me your feedbacks

**Comment by dixhuit** [comment]
Sounds promising!

**Comment by dixhuit** [comment]
@victorgarciaesgi It would be super useful for anyone in my position (still using Vuelidate and evaluating Regle) for the migration guide to include a section on how to migrate nested validation patterns from Vuelidate to Regle. I just checked the v0.7.4-beta.2 source and this isn't the case yet (understandable given it's still in beta). Do you plan to add this?

**Comment by victorgarciaesgi** [comment]
Yes I plan to do this but before merging, I'll do the docs when everything is finalised and tested

**Comment by NtchPlayer** [comment]
Wooo thanks you ! You add it so fast 😄 And thanks to update my reproduction to provide an example !

**Comment by victorgarciaesgi** [comment]
Great news! Nested component validation is available in `0.7.4`!!
https://github.com/victorgarciaesgi/regle/releases/tag/v0.7.4

**Comment by NtchPlayer** [comment]
Congralutation ! I try to add it quickly to my project !

**Comment by victorgarciaesgi** [comment]
Updated your reproduction with updated API, add the usage of `namespace`, which in your case allow to pass the namespace as a prop.
https://stackblitz.com/edit/vuelidate-nested-validation-example-qj8c39t1?file=app.vue
Docs:
https://reglejs.dev/advanced-usage/scoped-validation#namespaces-inside-scopes
